mavros_extras/src/plugins/landing_target.cpp

```diff
@@ -0,0 +1,242 @@
+		sp_nh.param<std::string>("mav_frame", mav_frame, "LOCAL_NED");
+
+		land_target_pub = sp_nh.advertise<geometry_msgs::PoseStamped>("landing_target", 10);
+		target_size_pub = sp_nh.advertise<geometry_msgs::Vector3>("target_size", 10);
+
+		if (listen_tf) {
+			ROS_INFO_STREAM_NAMED("landing_target", "Listen to landing_target transform " << tf_frame_id
+												      << " -> " << tf_child_frame_id);
+			tf2_start("LandingTargetTF", &LandingTargetPlugin::transform_cb);
+		}
+		else {
+			land_target_sub = sp_nh.subscribe("target", 10, &LandingTargetPlugin::land_target_cb, this);
+		}
+	}
+
+	const message_map get_rx_handlers() {
+		return {
+			       MESSAGE_HANDLER(MAVLINK_MSG_ID_LANDING_TARGET, &LandingTargetPlugin::handle_landing_target)
+		};
+	}
+
+private:
+	friend class TF2ListenerMixin;
+	ros::NodeHandle sp_nh;
+	UAS *uas;
+
+	bool send_tf;
+	bool listen_tf;
+	double tf_rate;
+	ros::Time last_transform_stamp;
+
+	std::string tf_frame_id;
+	std::string tf_child_frame_id;
+
+	ros::Publisher land_target_pub;
+	ros::Publisher target_size_pub;
+	ros::Subscriber land_target_sub;
+
+	double target_size_x, target_size_y;
+
+	std::string mav_frame;
+
+	/* -*- low-level send -*- */
+	void landing_target(uint64_t time_usec,
+			float angle_x,
+			float angle_y,
+			float distance,
+			float size_x,
+			float size_y,
+			uint8_t target_num,
+			uint8_t frame) {
+		mavlink_message_t msg;
+		mavlink_msg_landing_target_pack_chan(UAS_PACK_CHAN(uas), &msg,
+				time_usec,
+				angle_x,
+				angle_y,
+				distance,
+				size_x,
+				size_y,
+				target_num,
+				frame);
+		UAS_FCU(uas)->send_message(&msg);
+	}
+
+	/* -*- mid-level helpers -*- */
+
+	/*	AUXILIAR MATH: (for spherical coordinates)
+	 *
+	 *	d = sqrt(x^2 + y^2 + z^2)
+	 *	theta = atan(y / x)
+	 *	phi = atan(sqrt(x^2 + y^2) / z)
+	 *
+	 *	where,	theta	= angle_x	(note: notation in Mavlink/Firmware may lead to wrongly consider rectangular coordinates)
+	 *		phi	= angle_y
+	 *
+	 *	Conversion from spherical to rectangular coordinates:
+	 *	x = d * sin(phi) * cos(theta)
+	 *	y = d * sin(phi) * sin(theta)
+	 *	z = d * cos(phi)
+	 *
+	 */
+
+	/**
+	 * Send landing target transform to FCU
+	 */
+	void send_landing_target(const ros::Time &stamp, const Eigen::Affine3d &transf) {
+		// origin position in ROS ENU frame
+		auto pos = UAS::transform_frame_enu_ned(Eigen::Vector3d(transf.translation()));
+
+		float distance = Eigen::internal::psqrt(pos[0] * pos[0] + pos[1] * pos[1] + pos[2] * pos[2]);
+		float phi = atan(Eigen::internal::psqrt(pos[0] * pos[0] + pos[1] * pos[1]) / pos[2]);		// = angle_x
+		float theta = atan(pos[1] / pos[0]);		// = angle_y
+
+		float size_x_rad = target_size_x * phi;		// assuming this is the arc length of the circle in X-axis
+		float size_y_rad = target_size_y * theta;	// assuming this is the arc length of the circle in Y-axis
+
+		uint8_t frame = UAS::idx_frame(mav_frame);	// MAV_FRAME index based on given frame name
+		if (frame == -1) {
```

Comment:
**u**int8 can't be -1, also do finding once in int. mav_frame don't change over time, so why eat cpu cycles with useless calls?